File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/ListaSeleccionable.java
package com.victorlh.android.framework.listas.adapter.seleccionable;

import androidx.annotation.NonNull;

import com.victorlh.android.framework.listas.ItemLista;

import java.util.List;

public interface ListaSeleccionable<T extends ItemLista> {

	enum ETipoSeleccion {
		NONE, CLICK_UNICO, CLICK_MULTIPLE, LONGCLICK
	}

	void setTipoSeleccion(@NonNull ETipoSeleccion tipoSeleccion);

	ETipoSeleccion getTipoSeleccion();

	void setSeleccion(@NonNull T[] items);

	void setSeleccion(@NonNull List<T> items);

	void addSeleccion(@NonNull T item);

	void removeSeleccion(@NonNull T item);

	void removeAllSeleccion();

	boolean haySeleccionados();

	boolean isSeleccionado(@NonNull T item);

	T[] getItemsSeleccionados();

	void addOnStartSeleccionListener(@NonNull OnStartSeleccionListener onStartSeleccionListener);

	void addOnStopSeleccionListener(@NonNull OnStopSeleccionListener onStopSeleccionListener);

	void addOnSeleccionChangeListener(@NonNull OnSeleccionChangeListener onSeleccionChangeListener);

	void addOnItemSeleccionadoListener(@NonNull OnItemSeleccionadoListener<T> onItemSeleccionadoListener);

	void addOnItemDeseleccionadoListener(@NonNull OnItemDeseleccionadoListener<T> onItemDeseleccionadoListener);

	interface OnStartSeleccionListener {
		void onStartSeleccion();
	}

	interface OnStopSeleccionListener {
		void onStopSeleccion();
	}

	interface OnSeleccionChangeListener {
		void onSeleccionChangeListener(int numSeleccionados);
	}

	interface OnItemSeleccionadoListener<T extends ItemLista> {
		void onItemSeleccionado(T item);
	}

	interface OnItemDeseleccionadoListener<T extends ItemLista> {
		void onItemDeseleccionado(T item);
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/ListaSeleccionable.kt
package com.victorlh.android.framework.listas.adapter.seleccionable

interface ListaSeleccionable<T> {

    enum class ETipoSeleccion {
        NONE, CLICK_UNICO, CLICK_MULTIPLE, LONGCLICK
    }

    val tipoSeleccion: ETipoSeleccion
    val itemsSeleccionados: Array<T>

    fun setSeleccion(items: Array<T>)
    fun setSeleccion(items: List<T>)
    fun addSeleccion(item: T)
    fun removeSeleccion(item: T)
    fun removeAllSeleccion()
    fun haySeleccionados(): Boolean
    fun isSeleccionado(item: T): Boolean
    fun addOnStartSeleccionListener(onStartSeleccionListener: OnStartSeleccionListener)
    fun addOnStopSeleccionListener(onStopSeleccionListener: OnStopSeleccionListener)
    fun addOnSeleccionChangeListener(onSeleccionChangeListener: OnSeleccionChangeListener)
    fun addOnItemSeleccionadoListener(onItemSeleccionadoListener: OnItemSeleccionadoListener<T>)
    fun addOnItemDeseleccionadoListener(onItemDeseleccionadoListener: OnItemDeseleccionadoListener<T>)

    interface OnStartSeleccionListener {
        fun onStartSeleccion()
    }

    interface OnStopSeleccionListener {
        fun onStopSeleccion()
    }

    interface OnSeleccionChangeListener {
        fun onSeleccionChangeListener(numSeleccionados: Int)
    }

    interface OnItemSeleccionadoListener<T> {
        fun onItemSeleccionado(item: T)
    }

    interface OnItemDeseleccionadoListener<T> {
        fun onItemDeseleccionado(item: T)
    }
}